File: 208.实现Trie(前缀树).go
/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */

type Trie struct {
	is_end bool
	next [26]*Trie
}


func Constructor() Trie {
	return Trie{}
}


func (this *Trie) Insert(word string)  {
	node := this
	for _, c := range word {
		c -= 'a'
		if node.next[c] == nil {
			node.next[c] = &Trie()
		}
		node = node.next[c]
	}
	node.is_end = true
}


func (this *Trie) searchPrefix(word string) *Trie {
	node := this
	for _, c := range word {
		c -= 'a'
		if node.next[c] == nil {
			return nil
		}
		node = node.next[c]
	}
	return node
}


func (this *Trie) Search(word string) bool {
	node := searchPrefix(word)
	return node != nil && node.is_end
}


func (this *Trie) StartsWith(prefix string) bool {
	return this.searchPrefix(prefix) != nil
}

File: 208.实现Trie(前缀树).cpp
#include <string>

using std::string;

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
class Trie {
public:
    Trie(): is_end(false) {
        memset(next, 0, sizeof(next));
    }

    void insert(string word) {
        Trie* node = this;
        for (const char& c : word) {
            if (node->next[c - 'a'] == nullptr) {
                node->next[c - 'a'] = new Trie();
            }
            node = node->next[c - 'a'];
        }
        node->is_end = true;
    }

    bool search(string word) {
        Trie* node = this;
        for (const char& c : word) {
            node = node->next[c - 'a'];
            if (node == nullptr) {
                return false;
            }
        }
        return node->is_end;
    }

    bool startsWith(string prefix) {
        Trie* node = this;
        for (const char& c : prefix) {
            node = node->next[c - 'a'];
            if (node == nullptr) {
                return false;
            }
        }
        return true;
    }

private:
    bool is_end;
    Trie* next[26];
};
